content/blog/2021-12-07-ml-experiment-versioning.md

```diff
@@ -0,0 +1,219 @@
+still better for code. Tools like GitHub make distributed collaboration easy,
+and you can see incremental code changes. That's why experiments get split
+between Git for code and experiment tracking tools for meta-information (usually
+with a link in one or the other to keep track).
+
+ML experiment versioning combines experiment tracking and version control.
+Instead of managing these separately, keep everything in one place and get the
+benefits of both, like:
+
+- **Experiments as code**: Track meta-information in the repository and version
+  it like code.
+- **Versioned reproducibility**: Save and restore experiment state, and track
+  changes to only execute what's new.
+- **Distributed experiments**: Organize locally and choose what to share,
+  reusing your existing repo setup.
+
+![Experiment Versioning](2021-12-07/exp-versioning.png)
+
+# ML Experiments as Code
+
+Experiment versioning treats experiments as code. It saves all metrics,
+hyperparameters, and artifact information in text files that can be versioned by
+Git (DVC [data versioning](https://dvc.org/doc/start/data-and-model-versioning)
+backs up the artifacts themselves anywhere). You do not need a centralized
+database or online services. Git becomes a store for experiment
+meta-information.
+
+You can choose your own file formats and paths, which you can configure in DVC:
+
+```dvc
+$ dvc exp init -i
+This command will guide you to set up a default stage in dvc.yaml.
+See https://dvc.org/doc/user-guide/project-structure/pipelines-files.
+
+DVC assumes the following workspace structure:
+├── data
+├── metrics.json
+├── models
+├── params.yaml
+├── plots
+└── src
+
+Command to execute: python src/train.py
+Path to a code file/directory [src, n to omit]: src/train.py
+Path to a data file/directory [data, n to omit]: data/images/
+Path to a model file/directory [models, n to omit]:
+Path to a parameters file [params.yaml, n to omit]:
+Path to a metrics file [metrics.json, n to omit]:
+Path to a plots file/directory [plots, n to omit]: logs.csv
+```
+
+Once you set up your repo in this structure, you start to see the benefits of
+this approach. Experiment meta-information lives in readable files that are
+always available, and your code can stay clean. You can read, save, and version
+your meta-information:
+
+```dvc
+$ cat params.yaml
+train:
+  epochs: 10
+model:
+  conv_units: 128
+```
+
+```dvc
+$ cat metrics.json
+{"loss": 0.24310708045959473, "acc": 0.9182999730110168}
+```
+
+You can see what changed in parameters, code, or anything else:
+
+```dvc
+$ git diff HEAD~1 -- params.yaml
+diff --git a/params.yaml b/params.yaml
+index baad571a2..57d098495 100644
+--- a/params.yaml
++++ b/params.yaml
+@@ -1,5 +1,5 @@
+ train:
+   epochs: 10
+-model:
+-  conv_units: 16
++model:
++  conv_units: 128
+```
+
+With DVC, you can even compare lots of experiments from the terminal like you
+would in a dashboard:
+
+```dvctable
```

Comment:
I think `dvctable` doesn't expect the command `$ dvc exp show` at the top ... table should look quite different from what I see below on the screenshots

also, there is a way to colorize the header

Reply:
It actually works well! Those were before I added `dvctable`. Here it is now:

<img width="707" alt="Screen Shot 2021-12-03 at 5 20 44 PM" src="https://user-images.githubusercontent.com/2308172/144680599-14d0fa3c-bd73-4776-becc-bed7124aa8ef.png">